sbi/analysis/sensitivity_analysis.py

```diff
@@ -0,0 +1,497 @@
+
+        optimizer = optim.Adam(
+            list(self._regression_net.parameters()),
+            lr=learning_rate,
+        )
+        max_num_epochs = 2 ** 31 - 1 if max_num_epochs is None else max_num_epochs
+
+        # criterion / loss
+        criterion = MSELoss()
+
+        epoch, self._val_log_prob = 0, float("-Inf")
+        while epoch <= max_num_epochs and not self._converged(epoch, stop_after_epochs):
+            self._regression_net.train()
+            for parameters, observations in train_loader:
+                optimizer.zero_grad()
+                outputs = self._regression_net(parameters)
+                loss = criterion(outputs, observations)
+                loss.backward()
+                if clip_max_norm is not None:
+                    clip_grad_norm_(
+                        self._regression_net.parameters(),
+                        max_norm=clip_max_norm,
+                    )
+                optimizer.step()
+
+            epoch += 1
+
+            # calculate validation performance
+            self._regression_net.eval()
+
+            val_loss = 0.0
+            with torch.no_grad():
+                for parameters, observations in val_loader:
+                    outputs = self._regression_net(parameters)
+                    loss = criterion(outputs, observations)
+                    val_loss += loss.item()
+            self._val_log_prob = -val_loss / num_validation_examples
+            self._validation_log_probs.append(self._val_log_prob)
+
+            print("Training neural network. Epochs trained: ", epoch, end="\r")
+
+        return deepcopy(self._regression_net)
+
+    def _converged(self, epoch: int, stop_after_epochs: int) -> bool:
+        r"""
+        Return whether the training converged yet and save best model state so far.
+        Checks for improvement in validation performance over previous epochs.
+        Args:
+            epoch: Current epoch in training.
+            stop_after_epochs: How many fruitless epochs to let pass before stopping.
+        Returns:
+            Whether the training has stopped improving, i.e. has converged.
+        """
+        converged = False
+
+        posterior_nn = self._regression_net
+
+        # (Re)-start the epoch count with the first epoch or any improvement.
+        if epoch == 0 or self._val_log_prob > self._best_val_log_prob:
+            self._best_val_log_prob = self._val_log_prob
+            self._epochs_since_last_improvement = 0
+            self._best_model_state_dict = deepcopy(posterior_nn.state_dict())
+        else:
+            self._epochs_since_last_improvement += 1
+
+        # If no validation improvement over many epochs, stop training.
+        if self._epochs_since_last_improvement > stop_after_epochs - 1:
+            posterior_nn.load_state_dict(self._best_model_state_dict)
+            converged = True
+
+        return converged
+
+    def find_active(
```

Comment:
the function name is not really explained in the doc string. why is it called that way? Maybe call it `find_directions` or so?

Reply:
I agree, maybe that's a better name. I'll change it